analysis/aggregate/aggregate.go

```diff
@@ -0,0 +1,227 @@
+	}
+}
+
+func (m *Max) Result() int64 {
+	return m.max
+}
+
+func (m *Max) Reset() {
+	m.seenFirst = false
+	m.max = 0
+}
+
+// Min retains the minimum value in the aggregated data.
+type Min struct {
+	min       int64
+	seenFirst bool
+}
+
+func (m *Min) Add(n int64) {
+	if !m.seenFirst {
+		m.min = n
+		m.seenFirst = true
+		return
+	}
+	if m.min > n {
+		m.min = n
+	}
+}
+
+func (m *Min) Result() int64 {
+	return m.min
+}
+
+func (m *Min) Reset() {
+	m.seenFirst = false
+	m.min = 0
+}
+
+// Sum returns the sum of the aggregated data.
+type Sum struct {
+	sum int64
+}
+
+func (s *Sum) Add(n int64) {
+	s.sum += n
+}
+
+func (s *Sum) Result() int64 {
+	return s.sum
+}
+
+func (s *Sum) Reset() {
+	s.sum = 0
+}
+
+// Mean returns the arithmetic mean of the aggregated data.
+type Mean struct {
+	sum   int64
+	count int64
+}
+
+func (m *Mean) Add(n int64) {
+	m.count++
+	m.sum += n
+}
+
+func (m *Mean) Result() int64 {
+	if m.count == 0 {
+		return 0
+	}
+	return m.sum / m.count
+}
+
+func (m *Mean) Reset() {
+	m.sum = 0
+	m.count = 0
+}
+
+// Percentile returns the nth percentile sample from the aggregated data.
+type Percentile struct {
+	q float64
+	h *hdrhistogram.Histogram
+}
+
+func NewPercentile(quantile float64, maxValue int64) *Percentile {
+	return &Percentile{
+		q: quantile,
+		h: hdrhistogram.New(1, maxValue, 3),
+	}
+}
+
+func (p *Percentile) Add(n int64) {
+	err := p.h.RecordValue(n)
+	if err != nil {
+		// value too large, record as large as we can
+		p.h.RecordValue(p.h.HighestTrackableValue())
+	}
+}
+
+func (p *Percentile) Result() int64 {
+	v := p.h.ValueAtQuantile(p.q)
+	if v >= p.h.HighestTrackableValue() {
+		// make it very obvious that the actual value is unknown but large
+		return math.MaxInt64
+	}
+	return v
+}
+
+func (p *Percentile) Reset() {
+	p.h.Reset()
+}
+
+// IsValidAgg returns true if desc is a valid descriptor for an aggregator type.
+func IsValidAgg(desc string) bool {
+	switch desc {
+	case "max", "min", "mean", "avg", "sum":
+		return true
+
+	default:
+		if len(desc) >= 3 && desc[0] == 'p' {
+			_, err := strconv.Atoi(desc[1:])
+			if err == nil {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+// NewFromDescriptor returns an aggregator that implements desc.
+// Returns BadDescriptorError if desc cannot be parsed.
+func NewFromDescriptor(desc string) (Aggregator, error) {
+	f, err := NewFactoryFromDescriptor(desc)
+	if err != nil {
+		return nil, err
+	}
+	return f(), nil
+}
+
+// AggregatorFactory creates a new Aggregator initialized to zero.
+type AggregatorFactory func() Aggregator
+
+// NewFactoryFromDescriptor returns an AggregatorFactory that will create
+// Aggregators based on desc.  Returns BadDescriptorError if desc is not a valid descriptor.
+func NewFactoryFromDescriptor(desc string) (AggregatorFactory, error) {
+	switch desc {
+	case "max":
+		return func() Aggregator { return &Max{} }, nil
+
+	case "min":
+		return func() Aggregator { return &Min{} }, nil
+
+	case "mean", "avg":
```

Comment:
may want to remove "mean" as there is already "avg" -- having two that mean the same could be confusing